### DISCUSSION TITLE
Using GAMS as a solver within IDAES

### DISCUSSION THREAD

**Discussion by tk3016** [IDAES/idaes-pse#1411] [General]
As far as I understand, the initialization routine uses IPOPT as the default solver when initializing the units. Going through the documentation I figured out that I can use a different solver such as CONOPT or SCIP by using the solver argument. However, I run into an error when I try to use BARON/OCTERACT or even GAMS as the solver (the way it is used in Pyomo).
Is there a possibility to use these solvers for initializing unit models or even simulating or optimizing the flowsheets in IDAES?

**Comment by adowling2** [comment]
Hello @tk3016! Can you post the error message? Also, include the version of IDAES you are using.

**Comment by andrewlee94** [comment]
@tk3016 As @adowling2 said, we really need to see the error you are encountering to help you fix the issue.

The short answer is yes, you can use different solvers. However you should think a bit about which solver you want to use and pick one suited to the class of problem you are solving. E.g. BARON is a mixed-integer solver, and for initialization you generally should not have any integer components active in the system, thus BARON not be the best choice as it is focused on a different class of problems.

However, there are some caveats to this. The most notable caveat is that many IDAES models make use of "external functions" to handle some of the more complex non-linear calculations (notably root finding algorithms for cubic and Helmholtz equations of state). Not all solver interfaces support these, so if you are using these in your model then you will need to use a solver that supports these (anything using the ASL interface should work (which notably does not include BARON)).

Another caveat is that the current IDAES scaling tools are tailored narrowly to IPOPT, and do not work as expected with some other solvers. If you are using the `idaes.core.util.scaling.calculate_scaling_factors` function, then you should limit yourself to using IPOPT, or switch to doing scaling manually throguh Pyomo (we are working to fix that this year, but it won't be ready until November at the earliest).

**Comment by tk3016** [comment]
Dear Alex and Andrew,

Many thanks for your prompt response.
First of all my aplogies, I posted about GAMS in haste, there was an issue with my installation.
But my question still stands only for initialisation part, please see the screenshots below:
![image](https://github.com/IDAES/idaes-pse/assets/56822481/e6b07624-79ec-4e54-84ef-a70a17ff54ad)
In the screenshot above, I am trying to use SNOPT to initialise the unit model but I get a key error.
![image](https://github.com/IDAES/idaes-pse/assets/56822481/b5b29ebe-3ae3-49e8-bf2a-36281c98b510)
On double-checking, I have a working installation of SNOPT on my machine as seen in the screenshot above.

I am able to use solvers like IPOPT (easy because it is the default one), CONOPT, SCIP or even GAMS to initialise the unit. But I have seen that it doesn't work with some solvers such as SNOPT, KNITRO, etc.

The version of IDAES I am using is 2.4.
The version of Python I am using is 3.9.13

Please do not hesitate to let me know if you need any other information from my side.

Best wishes,
Tanuj

**Reply by andrewlee94** [reply]
@tk3016 I am not familiar with SNOPT, and unfortunately the error message you are getting was truncated so I cannot see the actual message.

A quick Google search however suggests that SNOPT can work in Pyomo but that you need to do some additional steps first: https://groups.google.com/g/snopt/c/pt9TL5kAGSI

To start, could you check that SNOPT works through Pyomo with a simple test problem. In theory IDAES will work with any compatible Pyomo solver, so the place to start is to make sure your solver works in Pyomo.
